httpbakery/form/client.go

```diff
@@ -0,0 +1,201 @@
+	// The Prompter to use to get the responses if this is nil then
+	// DefaultPrompter will be used.
+	Prompter Prompter
+}
+
+// Fill processes fields by first sorting them and then prompting for the
+// value of each one in turn.
+//
+// The fields are sorted by first by group name. Those in the same group
+// are sorted so that secret fields come after non-secret ones, finally
+// the fields are sorted by description.
+//
+// Each field will be prompted for, the returned value will then be
+// validated against the type for the field. If the returned value does
+// not validate correctly it will be prompted again twice before giving
+// up.
+//
+// If a field contains environment variables these will be checked in
+// order. The first one that is set will be used as the value of default
+// when Prompt is called.
+func (f PromptingFiller) Fill(fields environschema.Fields) (map[string]interface{}, error) {
+	fs := make(fieldSlice, len(fields))
+	var i int
+	for k, v := range fields {
+		fs[i] = field{
+			name:  k,
+			attrs: v,
+		}
+		i++
+	}
+	sort.Sort(fs)
+	form := make(map[string]interface{}, len(fields))
+	for _, field := range fs {
+		var err error
+		form[field.name], err = f.prompt(field.attrs)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot complete form")
+		}
+	}
+	return form, nil
+}
+
+type field struct {
+	name  string
+	attrs environschema.Attr
+}
+
+type fieldSlice []field
+
+func (s fieldSlice) Len() int {
+	return len(s)
+}
+
+func (s fieldSlice) Less(i, j int) bool {
+	a1 := s[i].attrs
+	a2 := s[j].attrs
+	if a1.Group != a2.Group {
+		return a1.Group < a2.Group
+	}
+	if a1.Secret != a2.Secret {
+		return a2.Secret
+	}
+	return a1.Description < a2.Description
+}
+
+func (s fieldSlice) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+
+func (f *PromptingFiller) prompt(attr environschema.Attr) (interface{}, error) {
+	prompter := f.Prompter
+	if prompter == nil {
+		prompter = DefaultPrompter
+	}
+	def := f.getDefault(attr)
+	for i := 0; i < 3; i++ {
+		val, err := prompter.Prompt(attr.Description, def, attr.Secret)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot get input")
+		}
+		switch attr.Type {
+		case environschema.Tbool:
+			b, err := schema.Bool().Coerce(val, nil)
+			if err == nil {
+				return b, nil
+			}
+		case environschema.Tint:
+			i, err := schema.Int().Coerce(val, nil)
+			if err == nil {
+				return i, nil
+			}
+		default:
+			return val, nil
+		}
+	}
+	return nil, errgo.New("too many invalid inputs")
+}
+
+func (f *PromptingFiller) getDefault(attr environschema.Attr) string {
+	if attr.EnvVar != "" {
+		if env := os.Getenv(attr.EnvVar); env != "" {
+			return env
+		}
+	}
+	for _, envVar := range attr.EnvVars {
+		if env := os.Getenv(envVar); env != "" {
+			return env
+		}
+	}
+	return ""
+}
+
+// A Prompter is used by a PromptingFiller to get the values of the
+// fields.
+type Prompter interface {
+	Prompt(description, def string, secret bool) (string, error)
```

Comment:
How about:

     // Prompt prompts for the value of the given named attribute.
     // It is always acceptable for an implementation to return a string,
     // which will be coerced to the appropriate type if possible.
     Prompt(name string, environschema.Attr) (interface{}, error)

Then an implementation can use as much or as little information
of the attribute as it likes to determine how to prompt, and we don't
need to worry about new fields being added to environschema.Attr
that might need to be passed to Prompt and thus break the Prompter interface.